velox/dwio/common/SelectiveRepeatedColumnReader.cpp

```diff
@@ -16,8 +16,120 @@
+      offsets_, rows.size(), &memoryPool_);
+  dwio::common::ensureCapacity<vector_size_t>(
+      sizes_, rows.size(), &memoryPool_);
+  auto rawOffsets = offsets_->asMutable<vector_size_t>();
+  auto rawSizes = sizes_->asMutable<vector_size_t>();
+  vector_size_t nestedLength = 0;
+  for (auto row : rows) {
+    if (!nulls || !bits::isBitNull(nulls, row)) {
+      nestedLength += allLengths_[row];
+    }
+  }
+  nestedRows_.resize(nestedLength);
+  vector_size_t currentRow = 0;
+  vector_size_t nestedRow = 0;
+  vector_size_t nestedOffset = 0;
+  for (auto rowIndex = 0; rowIndex < rows.size(); ++rowIndex) {
+    auto row = rows[rowIndex];
+    // Add up the lengths of non-null rows skipped since the last
+    // non-null.
+    for (auto i = currentRow; i < row; ++i) {
+      if (!nulls || !bits::isBitNull(nulls, i)) {
+        nestedOffset += allLengths_[i];
+      }
+    }
+    currentRow = row + 1;
+    // Check if parent is null after adding up the lengths leading
+    // up to this. If null, add a null to the result and keep
+    // looping. If the null is last, the lengths will all have been
+    // added up.
+    if (nulls && bits::isBitNull(nulls, row)) {
+      rawOffsets[rowIndex] = 0;
+      rawSizes[rowIndex] = 0;
+      bits::setNull(rawResultNulls_, rowIndex);
+      anyNulls_ = true;
+      continue;
+    }
+
+    auto lengthAtRow = allLengths_[row];
+    std::iota(
+        &nestedRows_[nestedRow],
+        &nestedRows_[nestedRow + lengthAtRow],
+        nestedOffset);
+    rawOffsets[rowIndex] = nestedRow;
+    rawSizes[rowIndex] = lengthAtRow;
+    nestedRow += lengthAtRow;
+    nestedOffset += lengthAtRow;
+  }
+  childTargetReadOffset_ += nestedOffset;
+}
+
+void SelectiveRepeatedColumnReader::compactOffsets(RowSet rows) {
+  auto rawOffsets = offsets_->asMutable<vector_size_t>();
+  auto rawSizes = sizes_->asMutable<vector_size_t>();
+  RowSet rowsToCompact;
+  if (valueRows_.empty()) {
+    valueRows_.resize(rows.size());
+    rowsToCompact = outputRows();
+  } else {
+    rowsToCompact = valueRows_;
+  }
+  if (rows.size() == rowsToCompact.size()) {
+    return;
+  }
+
+  int32_t current = 0;
+  bool moveNulls = shouldMoveNulls(rows);
+  for (int i = 0; i < rows.size(); ++i) {
+    auto row = rows[i];
+    while (rowsToCompact[current] < row) {
+      ++current;
+    }
+    VELOX_CHECK(rowsToCompact[current] == row);
+    valueRows_[i] = row;
+    rawOffsets[i] = rawOffsets[current];
+    rawSizes[i] = rawSizes[current];
+    if (moveNulls && i != current) {
+      bits::setBit(
+          rawResultNulls_, i, bits::isBitSet(rawResultNulls_, current));
+    }
+  }
+  numValues_ = rows.size();
+  valueRows_.resize(numValues_);
+  offsets_->setSize(numValues_ * sizeof(vector_size_t));
+  sizes_->setSize(numValues_ * sizeof(vector_size_t));
+}
+
+RowSet SelectiveRepeatedColumnReader::applyFilter(RowSet rows) {
+  if (!scanSpec_->filter()) {
+    return rows;
+  }
+  auto kind = scanSpec_->filter()->kind();
+  VELOX_CHECK(
```

Comment:
Should this be a DCHECK? Can we move this check somewhere where it runs once vs. once per batch of data?